plugins/none-ls/default.nix

```diff
@@ -2,287 +2,128 @@
-      `vim.diagnostic.config` for diagnostics to work as expected.
-      See `:help vim.diagnostic.config` for more info.
-    '';
-
-    sourcesItems = helpers.mkNullOrOption (
-      with types; listOf (attrsOf str)
-    ) "The list of sources to enable, should be strings of lua code. Don't use this directly";
   };
 
-  config = mkIf cfg.enable {
-    warnings = optional (cfg.enableLspFormat && (cfg.onAttach != null)) ''
-      You have enabled the lsp-format integration with none-ls.
-      However, you have provided a custom value to `plugins.none-ls.onAttach`.
-
-      -> The `enableLspFormat` option will thus not be able to add the `require('lsp-format').on_attach` snippet to `none-ls`.
-    '';
-
-    assertions = [
-      {
-        assertion = cfg.enableLspFormat -> config.plugins.lsp-format.enable;
-        message = ''
-          Nixvim: You have enabled the `lsp-format` integration with none-ls.
-          However, you have not enabled the `lsp-format` plugin itself (`plugins.lsp-format.enable = true`).
-        '';
-      }
-    ];
-
-    extraPlugins = [ cfg.package ];
-
-    extraConfigLua =
-      let
-        onAttach' =
-          if (cfg.onAttach == null) && cfg.enableLspFormat then
-            ''
-              require('lsp-format').on_attach
-            ''
-          else
-            cfg.onAttach;
-
-        setupOptions =
-          with cfg;
-          {
-            inherit
-              border
-              cmd
-              debounce
-              debug
-              ;
-            default_timeout = defaultTimeout;
-            diagnostic_config = diagnosticConfig;
-            diagnostics_format = diagnosticsFormat;
-            fallback_severity = fallbackSeverity;
-            log_level = logLevel;
-            notify_format = notifyFormat;
-            on_attach = helpers.mkRaw onAttach';
-            on_init = onInit;
-            on_exit = onExit;
-            root_dir = rootDir;
-            should_attach = shouldAttach;
-            temp_dir = tempDir;
-            update_in_insert = updateInInsert;
+  callSetup = false;
+  extraConfig =
+    cfg:
+    let
+      # Set a custom on_attach when enableLspFormat is enabled
+      # FIXME: Using `mkIf (...) (mkDefault "...")` would be better,
+      # but that'd make it difficult to implement the "has no effect" warning.
+      setupOptions =
+        cfg.settings
+        // optionalAttrs (cfg.enableLspFormat && cfg.settings.on_attach == null) {
+          on_attach.__raw = ''
+            require('lsp-format').on_attach
+          '';
+        };
+
+      # Summarize why lsp-format integration is enabled (used in warnings)
+      enableLspFormatStatus =
+        if options.plugins.none-ls.enableLspFormat.isDefined then
+          "You have enabled the lsp-format integration with none-ls."
+        else
+          "The lsp-format integration with none-ls was enabled automatically, because `plugins.lsp-format` is enabled.";
```

Comment:
Thinking about it, I'm not actually sure how `isDefined` handles option defaults. I have a feeling it'll always be true, since the _default_ counts as a definition.

We might be able to do `length definitions == 1`, assuming the default is still listed in `definitions` when an actual definition exists?

Reply:
Nope, just tested. `definitions` is always length 1, regardless of whether it's using the default value or an explicit definition.

I guess this `if` branch is pointless then. Seems a shame there's no way to test if an option is explicitly defined :thinking: